src/kubernetes.cc

```diff
@@ -811,22 +813,21 @@ void KubernetesReader::WatchMaster(
     }
     try {
       if (verbose) {
-        LOG(INFO) << "Locking completion mutex";
+        LOG(INFO) << "Resetting and running service";
       }
-      // A notification for watch completion.
-      std::mutex completion_mutex;
-      std::unique_lock<std::mutex> watch_completion(completion_mutex);
+      service->reset();
+      auto lifetime_thread = std::thread([&service]() { service->run(); });
       Watcher watcher(
```

Comment:
Notice how the thread is tied to the lifetime of the watcher, so you could have even named it `watcher_thread`. Extending that line of reasoning, you could have had the `Watcher` itself manage the thread, or even _be_ the thread. See if the following makes sense:
```c++
namespace {
static std::shared_ptr<boost::asio::io_service>& Reset(
    std::shared_ptr<boost::asio::service& service) {
  if (service->stopped()) {
    service->reset();
  }
  return service;
}

class Watcher : public std::thread {
 public:
  Watcher(const std::string& endpoint,
          std::function<void(json::value)> event_callback,
          std::shared_ptr<boost::asio::io_service> service, bool verbose)
      : std::thread([=]() { service->run(); }),
        name_("Watcher(" + endpoint + ")"), service_(service),
        event_parser_(event_callback), verbose_(verbose) {}
  ~Watcher() {}

  void operator()(const boost::iterator_range<const char*>& range,
                  const boost::system::error_code& error) {
    ...
    if (error) {
      ...
      service_->stop();
    }
  }

 private:
  std::string name_;
  std::shared_ptr<boost::asio::io_service> service_;
  json::Parser event_parser_;
  bool verbose_;
};
...
}  // namespace
```
and then:
```c++
    Watcher watcher(
        endpoint,
        [=](json::value raw_watch) {
          WatchEventCallback(callback, name, std::move(raw_watch));
        },
        Reset(service), verbose);
      http::client::response response = client.get(request, std::ref(watcher));
      if (verbose) {
        LOG(INFO) << "Waiting for completion";
      }
      watcher.join();
```

Reply:
Done. (Thanks!)